pkg/controller/machineset/machineset_controller_test.go

```diff
@@ -17,196 +17,134 @@ limitations under the License.
-				}
-
-				// TODO (robertbailey): Figure out why the control loop isn't working as expected.
-				/*
-					g.Eventually(func() int {
-						if err := c.List(context.TODO(), machines); err != nil {
-							return -1
-						}
-						return len(machines.Items)
-					}, timeout).Should(gomega.BeEquivalentTo(replicas))
-				*/
-			},
-		},
-	}
-
-	for _, tc := range testCases {
-		t.Logf("Running %q testcase", tc.name)
-		func() {
-			if err := c.Create(context.TODO(), tc.instance); err != nil {
-				t.Errorf("error creating instance: %v", err)
-			}
+		}
 
-			defer func() {
-				err := c.Delete(context.TODO(), tc.instance)
-				if err != nil {
-					t.Logf("error deleting %v: %v", tc.instance, err)
-				}
+		By("Creating the MachineSet")
+		Expect(k8sClient.Create(ctx, instance)).To(Succeed())
 
-				select {
-				case recv := <-requests:
-					if recv != tc.expectedRequest {
-						t.Error("received request does not match expected request")
-					}
+		machines := &machinev1.MachineList{}
 
-					t.Logf("Last reconcile %v", <-requests)
-				case <-time.After(timeout):
-					t.Error("timed out waiting for request")
-				}
-			}()
+		By("Verifying that we have 2 replicas")
+		Eventually(func() (int, error) {
+			if err := k8sClient.List(ctx, machines, client.InNamespace(namespace.Name)); err != nil {
+				return 0, err
+			}
+			return len(machines.Items), nil
+		}, timeout).Should(BeEquivalentTo(replicas))
+
+		By("Trying to delete 1 machine and check the MachineSet scales back up")
+		machineToBeDeleted := machines.Items[0]
+		Expect(k8sClient.Delete(ctx, &machineToBeDeleted)).To(Succeed())
+
+		// Verify that the Machine has been deleted.
+		Eventually(func() bool {
+			key := client.ObjectKey{Name: machineToBeDeleted.Name, Namespace: machineToBeDeleted.Namespace}
+			if err := k8sClient.Get(ctx, key, &machineToBeDeleted); apierrors.IsNotFound(err) || !machineToBeDeleted.DeletionTimestamp.IsZero() {
+				return true
+			}
+			return false
+		}, timeout).Should(BeTrue())
 
-			select {
-			case recv := <-requests:
-				if recv != tc.expectedRequest {
-					t.Errorf("received request does not match expected request. Got: %v, expected: %v", recv, tc.expectedRequest)
+		By("Verifying that we have 2 replicas")
+		Eventually(func() (ready int) {
+			if err := k8sClient.List(ctx, machines, client.InNamespace(namespace.Name)); err != nil {
+				return -1
+			}
+			for _, m := range machines.Items {
+				if !m.DeletionTimestamp.IsZero() {
+					continue
 				}
-			case <-time.After(timeout):
-				t.Error("timed out waiting for request")
+				ready++
 			}
-
-			tc.verifyFnc()
-		}()
+			return
+		}, timeout*3).Should(BeEquivalentTo(replicas))
+	})
+})
+
+func cleanResources() error {
+	machineSets := &machinev1.MachineSetList{}
+	if err := k8sClient.List(ctx, machineSets); err != nil {
+		return err
+	}
+	for _, machineSet := range machineSets.Items {
+		if err := k8sClient.Delete(ctx, &machineSet); err != nil {
+			return err
+		}
```

Comment:
Taking the address of a loop variable in go is asking for trouble! 😅 Do this instead to avoid confusion later 😉 
```suggestion
	for _, machineSet := range machineSets.Items {
	        ms := machineSet
		if err := k8sClient.Delete(ctx, &ms); err != nil {
			return err
		}
```
While I don't think it makes a problem here, it can in certain cases so it's best to just avoid it

Have a read of https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable for more info

Please also fix the one below